### PR TITLE
OCPBUGS-31298: Prevent no-operation patches from overloading the kube-api server with unnecessary calls

### PR DESCRIPTION
**Pull request by RishabhSaini**
Closes OCPBUGS-31298

**- What I did**
When the node status is updated, detect if `reconcileMaster` actually makes any changes to the node config. If empty then prevent making no-op calls

**- How to verify it**
1) #4277 increases the `nodeStatusReportFrequency` to 10s and adds logs when an empty patch is made in function `UpdateNodeRetry`.
  ```
  if bytes.Equal(patchBytes, []byte("{}")) {
      klog.Infof("No change detected in UpdateNodeRetry")
  }
  ```
  These logs will indicate reduction in the empty patch calls made to the api-server

2)  grepping the MCC's node update calls in the kube-api audit logs

**- Description for the changelog**

### CODE REVIEW

**Comment by RishabhSaini** [comment]
Currently failing on expecting empty patches in jsonmergepatch:

```
pkg/controller/container-runtime-config/container_runtime_config_controller_test.go:1652: expectPatchContainerRuntimeConfig(ctrc, []byte("{}"))

pkg/controller/kubelet-config/kubelet_config_controller_test.go:994: f.expectPatchKubeletConfig(kc, []byte("{}"))
```

**Review by mtrmac — COMMENTED** [review]
I’m not really qualified to review this, but, in general, this is unintuitive and surprising to me. It feels like a scattershot attempt to avoid the _effect_ of the problem but does not really address the _root cause_.

- If this were a truly widespread generic issue, I’d expect it to be handled at some generic level. The existing pattern of two `json.Marshal`+`Create…Patch`+`.Patch` is long enough; adding one more manual step to _every single_ operation is just not a way to reliably get things fixed without regressions. It costs too much to maintain all those copies, and at scale, there will certainly be instances where that check will be forgotten, or buggy.
- This is adding the check even in cases which are already trying to detect “nothing needs to change”; why is that code ineffective?
- Implementation notes, but really it seems to me that the general approach should be considered first:
  - Generating a JSON and comparing that is a pretty costly way to compare; wouldn’t it be possible to compare the Go data before marshaling to JSON?
  - So does the patch generate `"{}"` or `""`? I guess it’s only one; and if it really generates both, at the very least I’d expect the condition to be encapsulated in a function. …
  - … or, better, create a utility that combines all of the `Marshal(old); Marshal(new); `Create…Patch`; `.Patch` into a single function call.

Basically, either (unlikely?) this should happen in the generic k8s client libraries or generic operator utilities, or this instead should be a set of specific targeted fixes at individual sync loops so that they don’t try to trigger updates when not necessary.

**Comment by RishabhSaini** [comment]
Here is the investigation:
When I jsondiff the two nodes old and new, the only difference is in the condition field of the `status`. The `lastHeartbeatTime`  and `lastTransitionTime` time seem to update every `nodeStatusReportFrequency`, triggering an `UpdateFunc` from the `NodeInformer`. Every time any change occurs to the node ( `Add` , `Delete` , `Update` ) it tries to `reconcileMaster` which tries to switch the label on the node from "Worker" -> "NoSchedule". Since this label almost never changes every time the node status is reported, `strategicpatch.CreateTwoWayMergePatch` returns an empty {} . And this without being checked on server side makes an no-op patch to the node.

**Comment by RishabhSaini** [comment]
Proposed solutions:
1) Reconsider the bugfix to #1024 . Since `reconcileMaster` at most times triggers a no-op update
2) As this PR originally suggested: Check after creating a `CreateNWayMergePatch` whether the `patchBytes` are empty. If they are then don't trigger a k8s Patch call on the resource.
3) Modify the generic client APIs `strategicPatch` and `jsonmergepatch` to explicitly indicate empty `patchBytes` so that when consuming we can explicitly ignore these changes

**Comment by djoshy** [comment]
Good start (: 

In terms of strategy, maybe we could try using something like [reflect.DeepEqual](https://github.com/openshift/machine-config-operator/blob/217f60789684c1e696c8ac9e1f974bec631ebf5e/pkg/controller/machine-set-boot-image/machine_set_boot_image_controller.go#L544) before marshalling JSON/issuing patches. It compares the go structs directly so it is less expensive. We use this strategy in a few other places in the MCO. We probably don't want to log in those no-op cases(or at least bump the verbosity- I imagine we have several no-op patch events and we do not want the logs filled with them)

Also, in the case of the boot image controller; it only [initiates a patch](https://github.com/openshift/machine-config-operator/blob/217f60789684c1e696c8ac9e1f974bec631ebf5e/pkg/controller/machine-set-boot-image/machine_set_boot_image_controller.go#L467) if there is an update to the machinesets, so another check before the API call is not necessary.

Could you explain a little bit about the changes to the local script update? Most of the team does not use this script anymore, Zack's helpers, so any kind of context/comments here is helpful (:

**Comment by djoshy** [comment]
/retitle OCPBUGS-31298: Prevent no-operation patches from overloading the kube-api server with unnecessary calls

**Comment by RishabhSaini** [comment]
> Could you explain a little bit about the changes to the local script update? Most of the team does not use this script anymore,(Zack's helpers is preferred), so any kind of context/comments here is helpful (:

Agreed, the helper scripts in preference to modifying the hack scripts

**Review by djoshy — APPROVED** [review]
This looks good to me, I'll let @yuqi-zhang take a final look

nit: if you could update commit message to be a one liner, that would be great. Not blocking on this though (:

**Comment by yuqi-zhang** [comment]
/lgtm

Thanks for the patch and fixups! And thanks @djoshy for the reviews :)

The CI failures seem to be a system error so we can retry

/retest-required

**Pull request merged**
